File: src/leetcode/0003-Longest-Substring-Without-Repeating-Characters/solution.py
'''
@Author: shaqsnake
@Email: shaqsnake@gmail.com
@Date: 2019-05-09 13:47:55
@LastEditTime: 2019-08-13 15:10:37
@Description: 
'''


class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        d = {}
        # cur is the point to mark the first char that not duplicate before
        cur = res = 0
        for i, n in enumerate(s):
            if n in d and cur <= d[n]:
                # if find a duplicate char and the char is behind the cur, move cur next to this char
                cur = d[n] + 1
            else:
                # calculate the longest substring
                res = max(res, i - cur + 1)
            d[n] = i
        return res

    def lengthOfLongestSubstring2(self, s):
        """
        :type s: str
        :rtype: int
        """
        f = [0] * 256
        l = 0
        r = -1
        res = 0

        while l < len(s):
            if r + 1 < len(s) and f[ord(s[r + 1])] == 0:
                r += 1
                f[ord(s[r])] += 1
            else:
                f[ord(s[l])] -= 1
                l += 1

            res = max(res, r - l + 1)

        return res

File: src/leetcode/0003-Longest-Substring-Without-Repeating-Characters/solution.cpp
// Copyright (c) 2019 shaqsnake. All rights reserved.
/**
 * @Author: shaqsnake
 * @Email: shaqsnake@gmail.com
 * @Date: 2019-08-13 15:09:38
 * @LastEditTime: 2019-08-13 15:13:28
 * @Description: 3. Longest Substring Without Repeating Characters
 */

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> map;
        int res = 0;
        for (int i = 0, j = 0; i < s.size(); i++) {
            map[s[i]]++;
            while (map[s[i]] > 1) map[s[j++]]--;
            res = max(res, i - j + 1);
        }

        return res;
    }

    int lengthOfLongestSubstring2(string s) {
        unordered_set<char> set;
        int res = 0;
        for (int i = 0; i < s.size(); i++) {
            int j = i;
            while (j < s.size() && !set.count(s[j]))
                set.insert(s[j++]);
            res = max(res, j - i);
            set.clear();
        }
        
        return res;
    }
};
